### PR TITLE
gcc+nvptx: fix cuda driver linking once more

### PR DESCRIPTION
**Pull request by haampie**
Attempted fix for #18444 

GCC actually needs the CUDA driver lib (libcuda.so) instead of the runtime lib (libcudart.so).

This is an attempt to properly fix https://github.com/spack/spack/pull/17619 / https://github.com/spack/spack/pull/18000.

Currently I cannot really test it, since GCC +nvptx fails because of other reasons now :) they use sm_30 somewhere in their defaults, which my CUDA driver does not support.

--

I guess the `libs` property is special in spack, whereas my `driver_libs` property is custom. Can someone comment on whether this is a good idea or not? At any rate, there's two things: (a) we don't want the stub lib to end up in the rpaths, so we shouldn't add driver libs to `libs`, and (b) some packages actually need the driver lib.

### CODE REVIEW

**Comment by ax3l** [comment]
cc @baberlevi @matthiasdiener @michaelkuhn @ilbiondo @rblake-llnl for review

File: var/spack/repos/builtin/packages/cuda/package.py
Comment (rblake-llnl) [inline var/spack/repos/builtin/packages/cuda/package.py]: Do we also want to mask out 'stubs'?

**Comment by rblake-llnl** [comment]
I know that NVRTC needs access to the raw CUDA driver, so this functionality is needed. I just don't know the proper way to do this in spack-land.

- Do we need to add a variant to the cuda package so that downstream packages can specify that the need to link against libcuda? 
- Would it be better to find both libcuda and libcudart within the prefix, then mask out the necessary compatibility libraries?
- Would GCC be OK with using a stub libcuda if that doesn't get added to the rpath later?  
- Does spack even support linking against a libcuda stub without linking it to the rpath?

**Comment by haampie** [comment]
IIUC: the libcuda.so stub lib is useful when you want to compile on a machine that does not have the appropriate hardware / driver installed. If you do have the driver installed, libcuda.so is available as a system lib.

> Would it be better to find both libcuda and libcudart within the prefix, then mask out the necessary compatibility libraries?

I believe when you do this in the `def libs` function, it causes spack to set the rpath to the folder containing the stub library in the dependent package, which caused the following issue: https://github.com/spack/spack/issues/17607.

> Would GCC be OK with using a stub libcuda if that doesn't get added to the rpath later?

Seems it works: https://github.com/spack/spack/issues/18444#issuecomment-685423319.

> Does spack even support linking against a libcuda stub without linking it to the rpath?

Not sure what the question means? Do you mean to ask whether it's possible to remove certain paths from entering the rpath of a dependent package when they are part of the `spec.libs` property of the link dependency package? (I don't know the answer to that one.)

**Comment by rblake-llnl** [comment]
> > Does spack even support linking against a libcuda stub without linking it to the rpath?
> 
> Not sure what the question means? Do you mean to ask whether it's possible to remove certain paths from entering the rpath of a dependent package when they are part of the spec.libs property of the link dependency package? (I don't know the answer to that one.)

Yep, that's exactly what I was asking.

**Comment by ilbiondo** [comment]
As far as gcc+nvptx is concerned then it only needs the stubs library when GCC itself is built and this looks to be to link against `libgomp/plugin/cuda/cuda.h`. When compiling code targeted for GPUs with gcc all the PTX code is created without needing the Nvidia stack. As such the fix works for #18444  

The rpath question is interesting though as Nvidia clearly intend the stubs to be used for normal dynamic (LD_LIBRARY_PATH) linking and that the full libcuda.so will be located at runtime.

**Comment by haampie** [comment]
> Yep, that's exactly what I was asking.

Maybe if you specify it as a build dependency, not a link dependency.

**Comment by adamjstewart** [comment]
@svenevs

**Comment by svenevs** [comment]
Are we sure this isn't a GCC problem?  It seems to me like if this is an issue at build time then whatever they're doing for linking could also be the problem. (Emphasis: could)

Whether you're using the driver or runtime library is an application choice, a given package should have the right to use both (e.g. they produce two versions of their API, high and low level, in separate libraries). I'm not talking about a variant. Most packages will use the runtime api, but the road we're going down seems to imply only one or the other will be allowed which is invalid in my view. Both should be available, and the build system (not spack, the package build system like b2 or CMake) is responsible for selecting the right version.

CMake seems to just forward everything to nvidia's linker which appears to just magically know what to do. But this stuff is way out of my depth, I falsely approved the last PR...

**Comment by svenevs** [comment]
Put differently: maybe we should find a way to hack what's needed for GCC somehow but not use these changes generically. Only do things differently when gcc+nvptx is being built (???)

**Comment by rblake-llnl** [comment]
@svenevs , any code that uses NVRTC is going to have the same problem. Occasionally CUDA codes need to do things that aren't supported by the runtime.  When that happens they need to invoke the driver ABI.  Personally I have used cuLaunchKernel before, see here: https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXEC.html#group__CUDA__EXEC_1gb8f3dc3031b40da29d5f9a7139e52e15 .

When I use a routine like this, I need to link against a libcuda.so library or my application won't build. Ideally, I'd like my binary to pull in libcuda.so from LD_LIBRARY_PATH so it works on any system.

In my experience, the best option for applications would be to have the stubs directory available as a non-rpath library so codes can link against it, but have the runtime available as an rpath-linked library.

**Review by michaelkuhn — COMMENTED** [review]
I can only contribute a "works for us" since I am not really a CUDA expert. We could successfully build `gcc+nvptx@7.5.0` and `gcc+nvptx@9.3.0`, and use them to build/run some OpenMP offloading codes. We haven't tried anything else depending on CUDA, though.

**Comment by svenevs** [comment]
I'm going to reach out on the dev forums and see if any of the pros there know what we should actually be doing, the install layout seems to change depending on how it's installed (spack vs system package manager vs runfile) which seems to possibly be part of the problem (the layout also seems to change across cuda versions too). If we don't have an answer next week then let's just merge this and see if we break everything else or not (can revert quickly).

**Comment by haampie** [comment]
Any news here?

**Comment by scheibelp** [comment]
> I guess the libs property is special in spack, whereas my driver_libs property is custom. Can someone comment on whether this is a good idea or not?

This sounds correct given what you want to do: `libs` is automatically added to `RPATH` and included in `-L` for all dependents. It sounds like you only need these libraries when building a small set of packages, so 

>> Not sure what the question means? Do you mean to ask whether it's possible to remove certain paths from entering the rpath of a dependent package when they are part of the spec.libs property of the link dependency package? (I don't know the answer to that one.)

> Yep, that's exactly what I was asking.

Not at a fine-grained level: you can set `Package.transitive_rpaths = False` to only RPATH direct dependencies, but generally cannot customize RPATHs on a per-package basis. However, since only a few packages need `.driver_libs`, I think your PR handles this well.

I agree with https://github.com/spack/spack/pull/18449#discussion_r481832104 that the stub libraries should still be excluded from `.libs` - apologies if I missed discussion somewhere addressing that.

**Comment by rblake-llnl** [comment]
> I agree with [#18449 (comment)](https://github.com/spack/spack/pull/18449#discussion_r481832104) that the stub libraries should still be excluded from `.libs` - apologies if I missed discussion somewhere addressing that.

@scheibelp , I also got confused.  As written, the package only uses the .stub directory for `.driver_libs`, which I think _shouldn't_ mask out the stubs directory as that's the whole point.  The PR as stands LGTM.

**Comment by svenevs** [comment]
I'm so sorry, I completely spaced out on this.  I [just asked for help on the dev forums](https://forums.developer.nvidia.com/t/packaging-cuda-in-spack-rpath-vs-link-path-with-compat-stubs-driver-vs-runtime-apis/156926), I think we should wait for their response.  It doesn't seem right to keep `compat` and `stubs` in the runtime side (`libs` property, item `3`).  Please add to that question if I made a mistake / misunderstood, I tried my best to condense it into as short of a question as possible in the hopes that we can answer what to do :slightly_smiling_face:

**Comment by scheibelp** [comment]
>> I agree with #18449 (comment) that the stub libraries should still be excluded from .libs - apologies if I missed discussion somewhere addressing that.

> I also got confused. As written, the package only uses the .stub directory for .driver_libs, which I think shouldn't mask out the stubs directory as that's the whole point. The PR as stands LGTM.

Thank you for the clarification: I agree that `driver_libs` should include the stubs.

However, I think that the `.libs` method should still exclude the `stubs` and `compat` libraries as it did before. Or am I missing something there?

**Comment by svenevs** [comment]
I think @rblake-llnl is on the money about wanting `stubs` for the `driver_libs`, in fact that may actually be the only one desired there since `libcuda.so` IIUC is something we want to be able to change (e.g., user updates drivers and new one in `$LD_LIBRARY_PATH` or system libs folder should take effect).

> However, I think that the `.libs` method should still exclude the `stubs` and `compat` libraries as it did before. Or am I missing something there?

I think we want to mask them out here as well.  @haampie thank you so much for your relentless adventures here, I think what we should do is

1. Add masking out `compat` and `stubs` back into the `libs` code.  (make sure to re-test after adding this!)
2. Since this has been on hold for a long time (sorry...), and it's a very complicated situation, provided (1) works just merge it and if the NVIDIA folks send a response I'll ping here :slightly_smiling_face:

We want to avoid breaking the runtime api in order to get the driver api working :wink:

**Comment by haampie** [comment]
What do we do with this PR?

**Comment by adamjstewart** [comment]
I think there's a better way to do this. Look at the suite-sparse package for an example. We can define a single libs attribute that can be queried for specific libraries, like `spec['cuda:drivers'].libs`.

File: var/spack/repos/builtin/packages/cuda/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/cuda/package.py:182]: > What do we do with this PR?

I think the only thing needed for now it to restore the filtering of 'stubs' and 'compat' from `libs` (`driver_libs` should keep 'stubs' though as discussed before in this thread).

**Comment by scheibelp** [comment]
> I think there's a better way to do this. Look at the suite-sparse package for an example. We can define a single libs attribute that can be queried for specific libraries, like spec['cuda:drivers'].libs.

I think for now the question is what different sets of libraries to make visible. I think we can improve on the interface for selecting those sets in a future PR

**Comment by ax3l** [comment]
Just ping-ing @baberlevi @matthiasdiener @michaelkuhn @ilbiondo @rblake-llnl for finalizing this PR :)
(Sorry, not feeling like I can contribute to this one in particular, but will learn from your solution.)

**Comment by haampie** [comment]
@ax3l I just did what @adamjstewart said, so now gcc does `spec['cuda:cuda'].libs` to get `libcuda.so` and the rest of the world will use `spec['cuda'].libs` and it'll fetch `libcudart.so`. Or `spec['cuda:cudart,fft,blas'].libs` etc.

**Comment by adamjstewart** [comment]
Then we can get rid of the `driver_libs` stuff right?

**Comment by haampie** [comment]
Yes, fixed, this pr is done then :)

**Review by adamjstewart — APPROVED** [review]

**Comment by haampie** [comment]
Now for the driver libcuda, should we still skip searching the compat folder?

```
# find . -iname "*libcuda.so*" | xargs realpath | xargs du -sh
17M	/usr/local/cuda-10.2/compat/libcuda.so.440.118.02
48K	/usr/local/cuda-10.2/targets/x86_64-linux/lib/stubs/libcuda.so
```

Can somebody knowledgeable comment on this?

**Comment by maximumcats** [comment]
Axel pointed me at this thread, I will try to help :-)

> Now for the driver libcuda, should we still skip searching the compat folder?

I am not sure I fully understand the question but let me give an attempt. In any given CUDA install, the libcuda.so that ends up in compat/ should have the same ABI as the libcuda.so stub in $CUDA_PATH/lib64 and the default library that ends up in /usr/lib64. So if you already have one of these available at link time, you don't need the others.

The presence of the compat/libcuda.so is for cases where the version of the CUDA toolkit you have is newer than the installed (kernel-mode) driver. In that case, you would want to run with this compat version in your library path (at run time) rather than the default one that gets installed to /usr/lib64 with the CUDA driver. But it's nearly the same CUDA driver library, only with additional hooks that allows it to run on older (kernel-mode) drivers (and support for failing gracefully if you happened to use an API that just didn't exist on the older kernel-mode driver).

**Comment by aweits** [comment]
Informational: this change configures libgomp-plugin-nvptx.so to dlopen() libcuda.so rather than link it in, and does not require it to be present at build-time.

```diff
--- a/var/spack/repos/builtin/packages/gcc/package.py
+++ b/var/spack/repos/builtin/packages/gcc/package.py
@@ -521,10 +521,6 @@ def configure_args(self):
         # nvptx-none offloading for host compiler
         if spec.satisfies('+nvptx'):
             options.extend(['--enable-offload-targets=nvptx-none',
-                            '--with-cuda-driver-include={0}'.format(
-                                spec['cuda'].prefix.include),
-                            '--with-cuda-driver-lib={0}'.format(
-                                spec['cuda'].libs.directories[0]),
                             '--disable-bootstrap',
                             '--disable-multilib'])
 
@@ -561,10 +557,7 @@ def nvptx_install(self):
         options += ['--prefix={0}'.format(prefix)]
 
         options += [
-            '--with-cuda-driver-include={0}'.format(
-                spec['cuda'].prefix.include),
-            '--with-cuda-driver-lib={0}'.format(
-                spec['cuda'].libs.directories[0]),
+            '--without-cuda-driver',
         ]
 
         with working_dir('nvptx-tools'):
```

**Comment by ax3l** [comment]
@haampie do the last two comments help you to finalize the PR? :)

**Comment by markyoder** [comment]
Thanks folks! I am also struggling with a gcc+CUDA instal -- variations on something like:

`./spack install gcc+nvptx @10.2.0 `

Looking forward to this PR, and extremely appreciative for any details, best practices, and recommendations you can make after the PR is squared away. Thanks again for your extraordinary efforts!

**Comment by alecbcs** [comment]
@haampie @alalazo does it make sense to continue working on this PR or was this superseded by other work related to compilers as deps?

**Comment by alecbcs** [comment]
Closing with no response. Feel free to reopen if folks want to continue working on this.

**Pull request closed without merging**
